### PR TITLE
Remove left joins and use only one call to select method in post_view.rs

### PR DESCRIPTION
**Pull request by dullbananas**
This makes the code cleaner and might also improve build time because now there's only one call to `select`.

I will probably make this change to other views in separate pull requests in the future.

### CODE REVIEW

**Comment by dullbananas** [comment]
I will try to get rid of the other left joins

**Comment by phiresky** [comment]
This is great imo. I was just looking at those queries for #3872 and all those left joins are pretty hard to understand and reason about. This is much easier to read. It will need a bit of testing though to see if the query plans are still the same or better. I think they should be and could even be better because with left joins sometimes postgresql estimates multiple rows being retrieved from the left join.

**Comment by phiresky** [comment]
it would be great if you could also make it so all those person_id_join things aren't actually called when person_id is null (logged out) instead of calling it with a fake -1 value

**Comment by RocketDerp** [comment]
> it would be great if you could also make it so all those person_id_join things aren't actually called when person_id is null (logged out)

I don't understand the Rust code well enough to get any further, but I decided to clone the function/method and eliminate as much as I could that referenced the logged-in user id. Can someone please help me figure out how to eliminate the rest?

```
fn queries_anonymous<'a>() -> Queries<
  impl ReadFn<'a, PostView, (PostId, Option<PersonId>, bool)>,
  impl ListFn<'a, PostView, PostQuery<'a>>,
> {
  let is_creator_banned_from_community = exists(
    community_person_ban::table.filter(
      post_aggregates::community_id
        .eq(community_person_ban::community_id)
        .and(community_person_ban::person_id.eq(post_aggregates::creator_id)),
    ),
  );

// how dpo we eliminate these next 3 assignments, this is anonymous user, not needed
  let is_saved = |person_id_join| {
    exists(
      post_saved::table.filter(
        post_aggregates::post_id
          .eq(post_saved::post_id)
          .and(post_saved::person_id.eq(person_id_join)),
      ),
    )
  };

  let is_read = |person_id_join| {
    exists(
      post_read::table.filter(
        post_aggregates::post_id
          .eq(post_read::post_id)
          .and(post_read::person_id.eq(person_id_join)),
      ),
    )
  };

  let is_creator_blocked = |person_id_join| {
    exists(
      person_block::table.filter(
        post_aggregates::creator_id
          .eq(person_block::target_id)
          .and(person_block::person_id.eq(person_id_join)),
      ),
    )
  };

  let all_joins = move |query: post_aggregates::BoxedQuery<'a, Pg>, my_person_id: Option<PersonId>| {
    // The left join below will return None in this case
    let person_id_join = my_person_id.unwrap_or(PersonId(-1));

    query
      .inner_join(person::table)
      .inner_join(community::table)
      .inner_join(post::table)
// how do we eliminate these next 3 joins that are user/person references?
      .left_join(
        community_follower::table.on(
          post_aggregates::community_id
            .eq(community_follower::community_id)
        ),
      )
      .left_join(
        community_moderator::table.on(
          post::community_id
            .eq(community_moderator::community_id)
        ),
      )
      .left_join(
        post_like::table.on(
          post_aggregates::post_id
            .eq(post_like::post_id)
        ),
      )
      .left_join(
        person_post_aggregates::table.on(
          post_aggregates::post_id
            .eq(person_post_aggregates::post_id)
        ),
      )
      .select((
        post::all_columns,
        person::all_columns,
        community::all_columns,
        is_creator_banned_from_community,
        post_aggregates::all_columns,
        CommunityFollower::select_subscribed_type(),
// how do we eliminate these next 3 for anonymous?
        is_saved(person_id_join),
        is_read(person_id_join),
        is_creator_blocked(person_id_join),
        post_like::score.nullable(),
        coalesce(
          post_aggregates::comments.nullable() - person_post_aggregates::read_comments.nullable(),
          post_aggregates::comments,
        ),
      ))
  };

  let read =
    move |mut conn: DbConn<'a>,
          (post_id, my_person_id, is_mod_or_admin): (PostId, Option<PersonId>, bool)| async move {

      let mut query = all_joins(
        post_aggregates::table
          .filter(post_aggregates::post_id.eq(post_id))
          .into_boxed(),
        my_person_id,
      );

        query = query
          .filter(community::removed.eq(false))
          .filter(post::removed.eq(false))
          ;

      query.first::<PostViewTuple>(&mut conn).await
    };

  let list = move |mut conn: DbConn<'a>, options: PostQuery<'a>| async move {
    let person_id = options.local_user.map(|l| l.person.id);

    let mut query = all_joins(post_aggregates::table.into_boxed(), person_id);


      query = query
        .filter(community::deleted.eq(false))
        .filter(post::deleted.eq(false));


    // every SELECT has to labor away on removed filtering
      query = query
        .filter(community::removed.eq(false))
        .filter(post::removed.eq(false));

    if options.community_id.is_none() {
      query = query.then_order_by(post_aggregates::featured_local.desc());
    } else if let Some(community_id) = options.community_id {
      query = query
        .filter(post_aggregates::community_id.eq(community_id))
        .then_order_by(post_aggregates::featured_community.desc());
    }

    if let Some(creator_id) = options.creator_id {
      query = query.filter(post_aggregates::creator_id.eq(creator_id));
    }


    if let Some(url_search) = options.url_search {
      query = query.filter(post::url.eq(url_search));
    }

    if let Some(search_term) = options.search_term {
      let searcher = fuzzy_search(&search_term);
      query = query.filter(
        post::name
          .ilike(searcher.clone())
          .or(post::body.ilike(searcher)),
      );
    }

      query = query
        .filter(post::nsfw.eq(false))
        .filter(community::nsfw.eq(false));


    query = match options.sort.unwrap_or(SortType::Hot) {
      SortType::Active => query
        .then_order_by(post_aggregates::hot_rank_active.desc())
        .then_order_by(post_aggregates::published.desc()),
      SortType::Hot => query
        .then_order_by(post_aggregates::hot_rank.desc())
        .then_order_by(post_aggregates::published.desc()),
      SortType::Controversial => query.then_order_by(post_aggregates::controversy_rank.desc()),
      SortType::New => query.then_order_by(post_aggregates::published.desc()),
      SortType::Old => query.then_order_by(post_aggregates::published.asc()),
      SortType::NewComments => query.then_order_by(post_aggregates::newest_comment_time.desc()),
      SortType::MostComments => query
        .then_order_by(post_aggregates::comments.desc())
        .then_order_by(post_aggregates::published.desc()),
      SortType::TopAll => query
        .then_order_by(post_aggregates::score.desc())
        .then_order_by(post_aggregates::published.desc()),
      SortType::TopYear => query
        .filter(post_aggregates::published.gt(now - 1.years()))
        .then_order_by(post_aggregates::score.desc())
        .then_order_by(post_aggregates::published.desc()),
      SortType::TopMonth => query
        .filter(post_aggregates::published.gt(now - 1.months()))
        .then_order_by(post_aggregates::score.desc())
        .then_order_by(post_aggregates::published.desc()),
      SortType::TopWeek => query
        .filter(post_aggregates::published.gt(now - 1.weeks()))
        .then_order_by(post_aggregates::score.desc())
        .then_order_by(post_aggregates::published.desc()),
      SortType::TopDay => query
        .filter(post_aggregates::published.gt(now - 1.days()))
        .then_order_by(post_aggregates::score.desc())
        .then_order_by(post_aggregates::published.desc()),
      SortType::TopHour => query
        .filter(post_aggregates::published.gt(now - 1.hours()))
        .then_order_by(post_aggregates::score.desc())
        .then_order_by(post_aggregates::published.desc()),
      SortType::TopSixHour => query
        .filter(post_aggregates::published.gt(now - 6.hours()))
        .then_order_by(post_aggregates::score.desc())
        .then_order_by(post_aggregates::published.desc()),
      SortType::TopTwelveHour => query
        .filter(post_aggregates::published.gt(now - 12.hours()))
        .then_order_by(post_aggregates::score.desc())
        .then_order_by(post_aggregates::published.desc()),
      SortType::TopThreeMonths => query
        .filter(post_aggregates::published.gt(now - 3.months()))
        .then_order_by(post_aggregates::score.desc())
        .then_order_by(post_aggregates::published.desc()),
      SortType::TopSixMonths => query
        .filter(post_aggregates::published.gt(now - 6.months()))
        .then_order_by(post_aggregates::score.desc())
        .then_order_by(post_aggregates::published.desc()),
      SortType::TopNineMonths => query
        .filter(post_aggregates::published.gt(now - 9.months()))
        .then_order_by(post_aggregates::score.desc())
        .then_order_by(post_aggregates::published.desc()),
    };

    let (limit, offset) = limit_and_offset(options.page, options.limit)?;

    query = query.limit(limit).offset(offset);

    debug!("Post View Query: {:?}", debug_query::<Pg, _>(&query));

    query.load::<PostViewTuple>(&mut conn).await
  };

  Queries::new(read, list)
}
```

**Comment by dullbananas** [comment]
I will probably do it by making it select `false` for the booleans if there's no person

**Comment by RocketDerp** [comment]
Ok, I figured out that the entire Response object had obligations that the SQL had to meet. I ended up duplicating the entire source file and creating all new objects and even a new API path to better understand how the code works. It wasn't that bad overall, probably took me 30 minutes. I was able to strip out everything and anything that related to being logged-in. Now it is generating a SQL statement with the minimum of joins and I can better study the join performance issues.

**Comment by RocketDerp** [comment]
I see the most recent commits are blowing up. I am working with your commit from 30 hours ago, https://github.com/LemmyNet/lemmy/blob/815dd4333435d9476d6a9de999a6db6b861f4739/crates/db_views/src/post_view.rs

I ponder if it would make more sense to divorce the ReadFn from ListFn on this critical part of the site. We are talking fetching 1 single post vs. a list of posts, with that queries and ReadFn ListFn - you don't need all that sorting on a single post, and lots of other logic (is profile view isn't relevant to a single post, etc). I feel like it seeded the performance problems because the code is so difficult to restructure and analyze.

File: crates/db_views/src/post_view.rs
Comment (dessalines) [inline crates/db_views/src/post_view.rs:115]: Thanks, this is muuuch better.

**Review by dessalines — COMMENTED** [review]
Huge improvement, thanks. It probably wouldn't be too much work to do this for the `community_view.rs`, which only has `blocked` as its boolean, and possibly `comment_view.rs`, which has a few more, but its the same process as post_view. I leave it to you whether to add those to this PR, or do them separately.

**Comment by phiresky** [comment]
Could you maybe get this into a mergeable state? It's better if it's incomplete but working so we can merge the partial work instead of making everything perfect.

**Review by dessalines — APPROVED** [review]
@Nutomic and @phiresky should probably take a look before merging also.

**Comment by phiresky** [comment]
looks good to me 👍

File: crates/db_views/src/post_view.rs
Comment (Nutomic) [inline crates/db_views/src/post_view.rs:124]: You can probably write these as `sql_types::Bool(false)` or `false.into()`.

File: crates/db_views/src/post_view.rs
Comment (Nutomic) [inline crates/db_views/src/post_view.rs:127]: Is explicit type declaration necessary?

**Review by dessalines — APPROVED** [review]

**Comment by dessalines** [comment]
nutomic has some comments above too tho.

File: crates/db_views/src/post_view.rs
Comment (dullbananas) [reply crates/db_views/src/post_view.rs:112]: false

File: crates/db_views/src/post_view.rs
Comment (phiresky) [reply crates/db_views/src/post_view.rs:115]: likely. it's really hard to get rust to do any smart inference for these dynamic diesel things

Comment (dullbananas) [reply crates/db_views/src/post_view.rs:109]: I tried to remove `SqlType = sql_types::Bool` and a pile of disgusting error messages appeared

**Pull request merged**
